src/main/java/io/pinecone/PineconeIndexOperationClient.java

```diff
@@ -1,52 +1,77 @@
+        this.clientConfig = clientConfig;
+        this.url = url;
+    }
+
+    public PineconeIndexOperationClient(PineconeClientConfig clientConfig, OkHttpClient client) {
+        this(clientConfig, client, createUrl(clientConfig));
     }
 
     public PineconeIndexOperationClient(PineconeClientConfig clientConfig) {
-        this.clientConfig = clientConfig;
-        client = new DefaultAsyncHttpClient();
-        this.url = "https://controller." + clientConfig.getEnvironment() + ".pinecone.io/databases/";
+        this(clientConfig, new OkHttpClient());
+    }
+
+    private static String createUrl(PineconeClientConfig clientConfig) {
+        if (clientConfig.getApiKey() == null || clientConfig.getEnvironment() == null) {
+            throw new PineconeValidationException("Both API key and environment name are required for index operations.");
+        }
+
+        return BASE_URL_PREFIX + clientConfig.getEnvironment() + BASE_URL_SUFFIX;
     }
 
     public void deleteIndex(String indexName) throws IOException {
-        System.out.println("Sending delete index request:");
-        client.prepare("DELETE", url + indexName)
-                .setHeader("accept", "text/plain")
-                .setHeader("Api-Key", clientConfig.getApiKey())
-                .execute()
-                .toCompletableFuture()
-                .join();
-
-        client.close();
+        Request request = new Request.Builder()
+                .url(url + indexName)
+                .delete()
+                .addHeader(ACCEPT_HEADER, TEXT_PLAIN)
+                .addHeader(API_KEY, clientConfig.getApiKey())
+                .build();
+
+        try (Response response = client.newCall(request).execute()) {
+            if (!response.isSuccessful()) {
+                throw new IOException(response.message());
+            }
+        } finally {
+            close(client);
+        }
     }
 
     public void createIndex(CreateIndexRequest createIndexRequest) throws IOException {
-        client.prepare("POST", url)
-                .setHeader("accept", "text/plain")
-                .setHeader("content-type", "application/json")
-                .setHeader("Api-Key", clientConfig.getApiKey())
-                .setBody(createIndexRequest.toJson())
-                .execute()
-                .toCompletableFuture()
-                .join();
-
-        client.close();
-    }
-}
+        MediaType mediaType = MediaType.parse("application/json; charset=utf-8");
+        RequestBody requestBody = RequestBody.create(createIndexRequest.toJson(), mediaType);
 
+        Request request = new Request.Builder()
+                .url(url)
+                .post(requestBody)
+                .addHeader(ACCEPT_HEADER, TEXT_PLAIN)
+                .addHeader(CONTENT_TYPE, CONTENT_TYPE_JSON)
+                .addHeader(API_KEY, clientConfig.getApiKey())
+                .build();
 
+        try (Response response = client.newCall(request).execute()) {
```

Comment:
Note that the way I've designed currently is if the response is unsuccessful, i.e. if the index already exists or some invalid parameters caused server side exception, then the client will result in IOException with the cause of the message. Please let me know if you prefer raising PineconeException rather than IOException (same applies to delete functionality).

Reply:
Definitely `IOException` is too generic. I think the more specific we can make these exceptions the better. Take a look at how I'm approaching this in the [ts client](https://github.com/pinecone-io/pinecone-ts-client/blob/main/src/errors/http.ts#L96): 

```
export const mapHttpStatusError = (failedRequestInfo: FailedRequestInfo) => {
  switch (failedRequestInfo.status) {
    case 400:
      return new PineconeBadRequestError(failedRequestInfo);
    case 401:
      return new PineconeAuthorizationError(failedRequestInfo);
    case 404:
      return new PineconeNotFoundError(failedRequestInfo);
    case 500:
      return new PineconeInternalServerError(failedRequestInfo);
    case 501:
      return new PineconeNotImplementedError(failedRequestInfo);
    default:
      throw new PineconeUnmappedHttpError(failedRequestInfo);
  }
};
```

My recommendation would be to create a `HttpErrorMapper` class (or similar name) that knows how to map status codes to the appropriate error classes. Putting this logic into a different class allows us to have a nicer experience without duplicating logic for errors that can happen anywhere (e.g. 401 Authorization errors). And also greatly simplifies documentation when people come asking about what errors can occur.

Reply:
Thanks for providing me with the example, adding a mapper is the ideal goal, so I'll implement this change here.